source/manual/testing.html.md

```diff
@@ -0,0 +1,91 @@
+
+## GOV.UK End-to-end Tests
+
+Currently, we have one suite of end-to-end tests: the [Publishing End-to-end Tests](https://github.com/alphagov/publishing-e2e-tests#govuk-publishing-end-to-end-tests). This suite of tests is deprecated and [we have agreed to delete it soon](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md#delete-publishing-e2e-tests). The current tests are slow and brittle and do not run in a realistic GOV.UK environment.
+
+As GOV.UK shifts towards more instantaneous infrastructure, we should be able to create a new suite of tests that accurately represents a real GOV.UK environment. Until then we have agreed to rely on the other forms of testing described here.
+
+## Contract Tests
+
+Also known as "Pact Tests".
+
+Contract tests check that APIs exposed by one app (the "provider") are compatible with other apps that use those APIs ("consumers"). This is done by having a set of shared tests: the tests are run by the provider and the consumers.
+
+Apps that are automatically deployed all the way to Production should follow the [guidance on tests (“safety checks”) in the Continuous Deployment RFC](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md#safety-checks):
+
+- Each endpoint with multiple, internal consumer apps has at least one contract test.
+
+Contract tests are an important part of the overall test strategy for GOV.UK. If we visually add unit and integration tests, the result is a "chain" of test coverage:
+
+```
+Consumer (Unit tests) <---> API (Contract Test) <---> Provider (Unit tests) <---> ...
+```
+
+The chain can be brittle, though: it can't test incremental state changes across multiple apps - think about all the API calls and state changes involved in publishing a document on GOV.UK. We still want higher-level, end-to-end tests to check this.
+
+[Read more about how to write contract tests](https://docs.publishing.service.gov.uk/manual/pact-testing.html).
+
+## Unit, Integration, etc. Tests
+
+Most apps use the [RSpec](https://rspec.info/) test framework. Use this framework for all new apps. Older apps follow [the Rails guidance and use Minitest](https://guides.rubyonrails.org/testing.html#rails-meets-minitest). We have migrated some apps from Minitest to RSpec to support contract testing ([example](https://github.com/alphagov/collections/issues/2259)).
+
+Feature tests in RSpec should be written in a style consistent with this article: [How we write readable feature tests with rspec](https://about.futurelearn.com/blog/how-we-write-readable-feature-tests-with-rspec) ([example](https://github.com/alphagov/content-publisher/blob/3e2ad9d23852bbc691d52922283a33d2dbc9acb9/spec/features/editing_content/edit_edition_spec.rb)). Use RSpec for feature tests in all new apps. Some older apps use the [Cucumber](https://github.com/cucumber/cucumber-ruby) test framework for feature tests.
```

Comment:
These first two paragraphs overlap with the guidance provided in [conventions for rails applications](https://docs.publishing.service.gov.uk/manual/conventions-for-rails-applications.html) and seem to start from an immediate expectation you've got a Ruby app.

I'd suggest avoid replication and say "Most GOV.UK applications are built on Ruby and Rails and we have conventions for their testing approach"

As an aside I think with any phrasing around this we don't want to give suggestions or leave ambiguity that you might want to port an older app from minitest to RSpec or Cucumber to RSpec. As we seem to have had a few of these started with good intentions but then left in a long term bad halfway state - I recommend people don't start these things unless they're really sure they'll finish them.

Reply:
Oops, this is me not taking the time to read the document I linked to, sorry.

I've rewritten this section in https://github.com/alphagov/govuk-developer-docs/pull/3564/commits/ef4b156c0961b3f431fb2cbf539a7ddbd2ba6db1.

The only bit I'd like to keep is about Minitest, as I want to acknowledge this is part of how we test GOV.UK, but also that it's not the recommended way. Regarding your concern, collections is a good example where the migration was done in full. I can't find a current example of a half-way app - do you have one?

Reply:
Ugh I didn't realise. I've adjusted this in https://github.com/alphagov/govuk-developer-docs/pull/3564/commits/0c102c5b38d4e53486ce7ac8cc7ee2a29e720e56.

Regarding contract tests, I think I misremembered a [problem that we found around the same time as the work on contract tests in Collections](https://github.com/alphagov/frontend/pull/2472). I can see e.g. [Whitehall](https://github.com/alphagov/whitehall/blob/main/spec/service_consumers/pact_helper.rb) has both with apparently no issue.

How does it look now?

Reply:
Yeah - I don't quite understand what the situation with Collections was. I thought it was interplay with cucumber, pact and minitest but Whitehall is indeed the same. 

Yes looks good. I can see a minor issue that I'll open a separate comment for.